test/gcs/mock/mock.go

```diff
@@ -0,0 +1,421 @@
+package mock
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+)
+
+// I don't know if it is easier or not to use go mock, but we really only need two things:
+// 1) Ability to mimick creation of buckets and objects
+// 2) Ability to mimick returning errors
+//
+// We don't need arbitrary return values, so generators like go mock or testify might be
+// overkill and doesn't give us the flexibility we need (e.g., will have to specify and
+// and reason about the state after each call rather than just pretend we have this fake
+// storage. The behavior of these commands, at the level of detail we care about, is pretty
+// easy to replicate.
+
+var (
+	MethodNewStorageBucket    = Method("NewStorageBucket")
+	MethodDeleteStorageBucket = Method("NewDeleteStorageBucket")
+	MethodListChildrenFiles   = Method("ListChildrenFiles")
+	MethodListDirectChildren  = Method("ListDirectChildren")
+	MethodAttrObject          = Method("AttrObject")
+	MethodCopyObject          = Method("CopyObject")
+	MethodReadObject          = Method("ReadObject")
+	MethodWriteObject         = Method("WriteObject")
+	MethodDeleteObject        = Method("DeleteObject")
+	MethodDownload            = Method("Download")
+	MethodUpload              = Method("Upload")
+)
+
+// mock GCS Client
+type clientMocker struct {
+	// project with buckets
+	gcp map[project]*buckets
+	// error map
+	// - on each call of the higher level function that calls any number of methods
+	//	in this library, you can use SetError(map[Method]*ReturnError) or ClearError()
+	//	to create the error return values you want. Default is nil.
+	err map[Method]*ReturnError
```

Comment:
Not sure why not just return standard error

Reply:
If your calling function calls the method multiple times. Then if you want a knob to control number of times before an error is returned.